woocommerce/woocommerce
# Integrate JS remote logging package in WooCommerce Admin

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/49286.

This PR integrates the `@woocommerce/remote-logging` package in WooCommerce Admin to capture and log errors to the server.

Changes in WooCommerce Admin:

- The JS remote error logging package is added as a dependency in WooCommerce Admin.
- Add `isRemoteLoggingEnabled` to `window.wcSettings` to determine if remote logging is enabled (It also checks if the user has opted in for tracking.)
- The remote logger is initialized in the `init` function of the package with a rate limit of 1 minute when `isRemoteLoggingEnabled` is true.
- Capture and log errors using the `captureException` function in the error boundary component.

Additions in the remote logging package:

- Add `captureException` function to capture and log exceptions.
- Replace `window.wcTracks.isEnabled` with `getSetting( 'isRemoteLoggingEnabled' )` to determine if remote logging is enabled.
- Some documentation updates and naming changes.




### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

To test this PR, we need to use the local dev environment and follow the below steps to simulate the error and check if it is logged in the server.

1. Checkout this PR branch.
2. Run `pnpm install`
3. Run `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build`
4. Use local fresh site or delete `woocommerce_allow_tracking` option to test the changes.
5. Go to `WooCommerce > Home` page.
6. Open the browser dev tools and go to the console tab.
7. Run `window.localStorage.setItem( 'debug', 'wc:remote-logging')` to enable the debug logging.
8. Reload the page.
9. In the console, you should see the debug message `wc:remote-logging Remote logging is disabled.`
10. Install Code Snippets plugin and add a new snippet with the below code to enable the remote logging.

```php
update_option( 'woocommerce_feature_remote_logging_enabled', 'yes' );
update_option( 'woocommerce_allow_tracking', 'yes' );

add_filter(
	'option_woocommerce_remote_variant_assignment',
	function () {
		return 5;
	}
);
```

11. Reload the page, and you should see the debug message `wc:remote-logging RemoteLogger initialized`
12. Now, run `git cherry-pick 146a650205045904313aec7df7ff8cd3a1a2283b` (146a650205045904313aec7df7ff8cd3a1a2283b) to simulate the error.
13. Reload the page, and you should see the `wc:remote-logging Caught unhandled rejection` and `wc:remote-logging Skipping send error in development environment` messages in the console.
14. Stop pnpm command and run `pnpm --filter='@woocommerce/plugin-woocommerce' build` to build production files.
15. Reload the page, and you should see the `wc:remote-logging Caught unhandled rejection` and `wc:remote-logging Sending error to server` messages in the console.
16. Confirm the request is sent to the server by checking the network tab in the browser dev tools.
17. Run `git cherry-pick 701c795bf16874bd00b4739e871ef1da945ffa83` (701c795bf16874bd00b4739e871ef1da945ffa83) to simulate the error while rendering the component
18. Run `pnpm --filter='@woocommerce/plugin-woocommerce' build` again and wait for ~1 minute
19. Reload the page to see the `wc:remote-logging Logging error` and `wc:remote-logging Sending error to server` messages in the console.
20. Go to D156767-code to learn how to view the errors on the server. Confirm the errors are logged in the server.

![Screenshot 2024-07-31 at 14 37 16](https://github.com/user-attachments/assets/9573d4e8-4e03-4259-a3d0-ab5b79a1fc4f)

![Screenshot 2024-07-31 at 14 39 53](https://github.com/user-attachments/assets/948034bc-5742-4369-9345-8784197b1ce8)




### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
